packages/cli/src/controller/publish-controller.ts

```diff
@@ -0,0 +1,65 @@
+// Copyright 2020-2021 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import fs from 'fs';
+import path from 'path';
+import {loadProjectManifest} from '@subql/common';
+import IPFS from 'ipfs-http-client';
+
+// https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#filecontent
+type FileContent = Uint8Array | string | Iterable<Uint8Array> | Iterable<number> | AsyncIterable<Uint8Array>;
+
+// https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#fileobject
+type FileObject = {
+  path?: string;
+  content?: FileContent;
+  mode?: number | string;
+  mtime?: Date | number[] | {secs: number; nsecs?: number};
+};
+
+type IterableType<T> = T extends AsyncIterable<infer U> ? U : T extends Iterable<infer U> ? U : T;
+
+type AddResult = IterableType<ReturnType<IPFS.IPFSHTTPClient['addAll']>>;
+
+export async function uploadToIpfs(ipfsEndpoint: string, projectDir: string): Promise<AddResult[]> {
+  const ipfs = IPFS.create({url: ipfsEndpoint});
+
+  const files = getFiles(projectDir);
+
+  const pendingResults = ipfs.addAll(files, {
+    wrapWithDirectory: true,
+    pin: true, // Theres no guarantee the server will respect this
+  });
+
+  const results: AddResult[] = [];
+
+  for await (const res of pendingResults) {
+    results.push(res);
+  }
+
+  return results;
+}
+
+function getFiles(projectDir: string): FileObject[] {
+  const projectManifestPath = path.resolve(projectDir, 'project.yaml');
```

Comment:
Is `.yml` a supported extension?

Reply:
.Yml is supported by the lib, though I think we can stick to .yaml at the moment. To avoid extra work for the hosted service